builtin/providers/azurerm/provider.go

```diff
@@ -0,0 +1,87 @@
+package azurerm
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+// Provider returns a terraform.ResourceProvider.
+func Provider() terraform.ResourceProvider {
+	return &schema.Provider{
+		Schema: map[string]*schema.Schema{
+			"subscription_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("ARM_SUBSCRIPTION_ID", ""),
+			},
+
+			"client_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("ARM_CLIENT_ID", ""),
+			},
+
+			"client_secret": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("ARM_CLIENT_SECRET", ""),
+			},
+
+			"tenant_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("ARM_TENANT_ID", ""),
+			},
+		},
+
+		ResourcesMap: map[string]*schema.Resource{
+			"azurerm_resource_group":  resourceArmResourceGroup(),
+			"azurerm_virtual_network": resourceArmVirtualNetwork(),
+		},
+
+		ConfigureFunc: providerConfigure,
+	}
+}
+
+// Config is the configuration structure used to instantiate a
+// new Azure management client.
+type Config struct {
+	ManagementURL string
+
+	SubscriptionID string
+	ClientID       string
+	ClientSecret   string
+	TenantID       string
+}
+
+const noConfigError = `Credentials must be provided either via subscription_id, client_id,
+client_secret and tenant_id. Please see the provider documentation for
+more information on how to obtain these credentials.`
+
+func providerConfigure(d *schema.ResourceData) (interface{}, error) {
+	config := Config{
+		SubscriptionID: d.Get("subscription_id").(string),
+		ClientID:       d.Get("client_id").(string),
+		ClientSecret:   d.Get("client_secret").(string),
+		TenantID:       d.Get("tenant_id").(string),
+	}
+
+	if config.SubscriptionID == "" || config.ClientID == "" || config.ClientSecret == "" || config.TenantID == "" {
+		return nil, fmt.Errorf(noConfigError)
+	}
```

Comment:
This is no longer coherent now that there's no config-by-file, yeah? We can just let the schema validate the presence of the four required values?